i3lock.c

```diff
@@ -645,6 +645,120 @@ void handle_screen_resize(void) {
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
+        }
+    }
+
+    free(buf);
+    return count;
+}
+
+static cairo_surface_t *read_raw_image(const char *image_path, const char *image_raw_format) {
+    cairo_surface_t *img;
+
+    /* Parse format as <width>x<height>[:<format>] */
+    int w = -1, h = -1;
+    char format[] = "native";
+    char *fmtstring = NULL;
+    if (asprintf(&fmtstring, "%%ix%%i:%%%lds", sizeof(format) - 1) == -1) {
+        errx(EXIT_FAILURE, "asprintf() failed.");
+    }
+
+    if (sscanf(image_raw_format, fmtstring, &w, &h, format) == EOF || w == -1 || h == -1) {
+        fprintf(stderr, "Invalid image format \"%s\"\n", image_raw_format);
+        free(fmtstring);
+        return NULL;
+    }
+
+    free(fmtstring);
+
+    /* Create image surface */
+    img = cairo_image_surface_create(CAIRO_FORMAT_RGB24, w, h);
+    if (cairo_surface_status(img) != CAIRO_STATUS_SUCCESS) {
+        fprintf(stderr, "Could not create surface: %s\n",
+                cairo_status_to_string(cairo_surface_status(img)));
+        return NULL;
+    }
+    cairo_surface_flush(img);
+
+    /* Use uint32_t* because cairo uses native endianness */
+    uint32_t *data = (uint32_t *)cairo_image_surface_get_data(img);
+    const int stride = cairo_image_surface_get_stride(img) / 4;
+
+    FILE *f = fopen(image_path, "r");
+    if (f == NULL) {
+        fprintf(stderr, "Could not open image \"%s\": %s\n",
+                image_path, strerror(errno));
+        cairo_surface_destroy(img);
+        return NULL;
+    }
+
+    /* Read the image, respecting cairo's stride, according to the format */
+    int size, count;
+    if (strcmp(format, "native") == 0) {
+        /* If the format is 'native', just read each line directly into the buffer */
+        size = w * h * 4;
+        count = read_raw_image_native(data, f, w, h, stride);
+    } else if (strcmp(format, "rgb") == 0) {
+        /* If the format is 'rgb', we have to convert it to argb */
+        size = w * h * 3;
+        count = read_raw_image_rgb(data, f, w, h, stride);
+    } else {
+        fprintf(stderr, "Unknown raw format: %s\n", format);
+        fclose(f);
+        cairo_surface_destroy(img);
+        return NULL;
+    }
+    cairo_surface_mark_dirty(img);
+
+    if (count < size) {
+        if (count < 0 || ferror(f)) {
+            fprintf(stderr, "Failed to read image \"%s\": %s\n",
+                    image_path, strerror(errno));
+            fclose(f);
+            cairo_surface_destroy(img);
+            return NULL;
+        } else {
+            /* Print a warning if the file contains less data than expected,
+             * but don't abort. It's useful to see how the image looks even if it's wrong. */
+            fprintf(stderr, "Warning: expected to read %i bytes from \"%s\", read %i\n",
+                    size, image_path, count);
+        }
+    }
+
+    fclose(f);
```

Comment:
I didn't check if any of the error paths leak something. This is complicated enough that I would want to double check, but I am tired enough that I do not want to do this. I'll just assume someone double checked this.

Reply:
All returns are preceded by an fclose(f), so that looks good to me.